core-runner/src/runner.rs

```diff
@@ -153,17 +171,18 @@ impl RunNextResult {
 pub struct Runner<SC: StorageCarrier, E: Environment<Ext>> {
     pub(crate) program_storage: SC::PS,
     pub(crate) message_queue: SC::MQ,
-    pub(crate) wait_list: SC::WL,
     pub(crate) log: Log,
     pub(crate) config: Config,
     env: E,
     block_info: BlockInfo,
+    wait_list: BTreeMap<Vec<u8>, MessageDispatch>,
```

Comment:
```suggestion
    wait_list: BTreeMap<(ProgramId, MessageId), MessageDispatch>,
```

at least

not sure why you need `ProgramId` though, `MessageId` is unique

Reply:
I've changed the key type.